### PR TITLE
Fix issues with concurrent `helm repo add` commands

### PR DESCRIPTION
**Pull request by pdecat**
**What this PR does / why we need it**:

This PR aims to fix the issue with concurrent `helm repo add` commands reported in #5677.
It adds a test case to reproduce the issue:

```
# go test -v ./cmd/helm/ -run=TestRepoAdd
=== RUN   TestRepoAddCmd
=== RUN   TestRepoAddCmd/add_a_repository
--- PASS: TestRepoAddCmd (0.01s)
    helm_test.go:154: $HELM_HOME has been configured at .
    --- PASS: TestRepoAddCmd/add_a_repository (0.01s)
=== RUN   TestRepoAdd
--- FAIL: TestRepoAdd (0.01s)
    helm_test.go:154: $HELM_HOME has been configured at .
    repo_add_test.go:106: test-name-1 was not successfully inserted into /tmp/helm-repotest-391309255/repository/repositories.yaml
    repo_add_test.go:106: test-name-2 was not successfully inserted into /tmp/helm-repotest-391309255/repository/repositories.yaml
FAIL
FAIL    k8s.io/helm/cmd/helm    0.069s
```
```
# go test -v ./cmd/helm/ -run=TestRepoAdd
=== RUN   TestRepoAddCmd
=== RUN   TestRepoAddCmd/add_a_repository
--- PASS: TestRepoAddCmd (0.01s)
    helm_test.go:154: $HELM_HOME has been configured at .
    --- PASS: TestRepoAddCmd/add_a_repository (0.00s)
=== RUN   TestRepoAdd
--- FAIL: TestRepoAdd (0.01s)
    helm_test.go:154: $HELM_HOME has been configured at .
    repo_add_test.go:106: test-name-0 was not successfully inserted into /tmp/helm-repotest-560212793/repository/repositories.yaml
    repo_add_test.go:106: test-name-2 was not successfully inserted into /tmp/helm-repotest-560212793/repository/repositories.yaml
FAIL
FAIL    k8s.io/helm/cmd/helm    0.067s
```

**Special notes for your reviewer**:

The fix is implemented with github.com/gofrs/flock for cross platform file locking.
Timeout is hard-coded to 5 minutes for now.

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by pdecat** [comment]
I've implemented file locking to achieve cross processes synchronization.
This makes the manual testing work fine:
```
# helm repo list                                                                                                                                                                        
Error: no repositories to show                                                                                                                                                           
# helm repo add repo1 https://kubernetes-charts.storage.googleapis.com & helm repo add repo2 https://kubernetes-charts.storage.googleapis.com &                                         
[1] 3604                                                                                                                                                                                 
[2] 3605                                                                                                                                                                                 
# wait $(jobs -p)                                                                                                                                                                       
"repo2" has been added to your repositories                                                                                                                                              
"repo1" has been added to your repositories                                                                                                                                              
[1]-  Done                    helm repo add repo1 https://kubernetes-charts.storage.googleapis.com                                                                                       
[2]+  Done                    helm repo add repo2 https://kubernetes-charts.storage.googleapis.com                                                                                       
# helm repo list                                                                                                                                                                        
NAME    URL                                                                                                                                                                              
repo2   https://kubernetes-charts.storage.googleapis.com                                                                                                                                 
repo1   https://kubernetes-charts.storage.googleapis.com                                                                                                                                 
```

However, it does not make the unit test pass as file locks are bound to processes, not file descriptors.
The test needs to be updated to perform repo additions in sub-processes.

**Comment by pdecat** [comment]
Implementation updated to use github.com/gofrs/flock for cross platform file locking.
This library also enables same process go routines synchronization.

I've updated unit tests to test both same process and concurrent processes repository additions.

File: cmd/helm/repo_add.go
Comment (mattfarina) [inline cmd/helm/repo_add.go]: I'm wondering if `5*time.Minute` is too long to wait. I've never seen the operation take more than a few seconds. Is there a reason for such a long time?

File: cmd/helm/repo_add.go
Comment (mattfarina) [inline cmd/helm/repo_add.go]: It might be worth putting a comment on why `LoadRepositoriesFile` is being run a second time. I see it because I read another conversation on it recently. May not be obvious to everyone or even me in the future.

File: glide.yaml
Comment (mattfarina) [inline glide.yaml]: Can you copy comment over to the new version of the file

File: glide.yaml
Comment (mattfarina) [inline glide.yaml]: Can you copy `testImports` over to the updated version of the file

File: cmd/helm/repo_add_test.go
Comment (mattfarina) [inline cmd/helm/repo_add_test.go]: Are the file locks bound to processes, threads, goroutines, or something else? goroutines are scheduled on top of threads and not guaranteed to be on different ones.

**Review by mattfarina — CHANGES_REQUESTED** [review]
Thanks for taking a crack at this. I didn't realize this library existed. Would have saved me some trouble in the past.

File: cmd/helm/repo_add.go
Comment (pdecat) [reply cmd/helm/repo_add.go]: I've never seen it take so much time either. It shouldn't happen unless another process is holding a lock on the file for that long. Or the file had grown in size ludicrously.

I only choose 5 minutes to make sure it is not aborted too early.
Maybe there could be a parameter to allow changing this from the CLI with a way lower default that could be raised in case of issues.

File: cmd/helm/repo_add_test.go
Comment (pdecat) [reply cmd/helm/repo_add_test.go]: This comment is obsolete, it comes from my first unix only implementation: https://github.com/helm/helm/pull/5678/commits/7ce22a3f585b574a2503de93a3f80ec75c1e5afb

github.com/gofrs/flock takes care of in process goroutine synchronization using mutexes in addition to using system level locks on files.

=> comment to remove.

**Comment by pdecat** [comment]
Rebased on master.

File: cmd/helm/repo_add.go
Comment (pdecat) [reply cmd/helm/repo_add.go]: I've lowered it to 30 seconds for now.

**Comment by pdecat** [comment]
I had to decrease the number of concurrent sub-processes from 3 to 2 to for the `helm repo add` test case to fit within CircleCI's pipeline memory limit.

Otherwise, the job would fail randomly during unit tests execution with the following error:

```
/usr/local/go/pkg/tool/linux_amd64/link: signal: killed
```

**Comment by pdecat** [comment]
@mattfarina I believe this PR is ready for final review.

**Comment by pdecat** [comment]
Rebased on master to fix conflicts on `glide.yaml`.

**Review by mattfarina — APPROVED** [review]
@pdecat Thanks. This looks good. FYI, It requires two approvals because it's a large code change.

**Comment by pdecat** [comment]
Hi @bacongobbler and @hickeyma, as the most active maintainers here apart from @mattfarina who already reviewed and approved this PR, could you please take a look at it?

**Review by hickeyma — APPROVED** [review]
LGTM, thanks.

**Pull request merged**
